File: app/typescript/src/domain/supplier/supplier.repository.ts
import { PrismaClient } from '@prisma/client'

// 取引先区分
export const SupplierType = {
  CUSTOMER: 'CUSTOMER', // 得意先
  VENDOR: 'VENDOR', // 仕入先
  BOTH: 'BOTH', // 得意先・仕入先両方
} as const

export type SupplierTypeValue = (typeof SupplierType)[keyof typeof SupplierType]

// 取引先の入力型
export interface CreateSupplierInput {
  supplierCode: string
  supplierName: string
  supplierNameKana?: string
  supplierType: SupplierTypeValue
  postalCode?: string
  address?: string
  phone?: string
  fax?: string
  email?: string
}

// 取引先の出力型
export interface Supplier {
  supplierCode: string
  supplierName: string
  supplierNameKana: string | null
  supplierType: string
  postalCode: string | null
  address: string | null
  phone: string | null
  fax: string | null
  email: string | null
  updatedAt: Date
}

export class SupplierRepository {
  constructor(private prisma: PrismaClient) {}

  async create(input: CreateSupplierInput): Promise<Supplier> {
    const result = await this.prisma.$queryRaw<Supplier[]>`
      INSERT INTO suppliers (
        supplier_code, supplier_name, supplier_name_kana, supplier_type,
        postal_code, address, phone, fax, email
      )
      VALUES (
        ${input.supplierCode}, ${input.supplierName}, ${input.supplierNameKana ?? null},
        ${input.supplierType}::supplier_type,
        ${input.postalCode ?? null}, ${input.address ?? null},
        ${input.phone ?? null}, ${input.fax ?? null}, ${input.email ?? null}
      )
      RETURNING
        supplier_code as "supplierCode",
        supplier_name as "supplierName",
        supplier_name_kana as "supplierNameKana",
        supplier_type as "supplierType",
        postal_code as "postalCode",
        address,
        phone,
        fax,
        email,
        updated_at as "updatedAt"
    `
    return result[0]
  }

  async findByCode(supplierCode: string): Promise<Supplier | null> {
    const result = await this.prisma.$queryRaw<Supplier[]>`
      SELECT
        supplier_code as "supplierCode",
        supplier_name as "supplierName",
        supplier_name_kana as "supplierNameKana",
        supplier_type as "supplierType",
        postal_code as "postalCode",
        address,
        phone,
        fax,
        email,
        updated_at as "updatedAt"
      FROM suppliers
      WHERE supplier_code = ${supplierCode}
    `
    return result.length > 0 ? result[0] : null
  }

  async findByType(supplierType: SupplierTypeValue): Promise<Supplier[]> {
    const result = await this.prisma.$queryRaw<Supplier[]>`
      SELECT
        supplier_code as "supplierCode",
        supplier_name as "supplierName",
        supplier_name_kana as "supplierNameKana",
        supplier_type as "supplierType",
        postal_code as "postalCode",
        address,
        phone,
        fax,
        email,
        updated_at as "updatedAt"
      FROM suppliers
      WHERE supplier_type = ${supplierType}::supplier_type
        OR supplier_type = 'BOTH'::supplier_type
      ORDER BY supplier_code
    `
    return result
  }
}

File: app/csharp/src/ProductionManagement.Infrastructure/Persistence/Repositories/SupplierRepository.cs
using Dapper;
using Npgsql;
using ProductionManagement.Application.Port.Out;
using ProductionManagement.Domain.Models.Supplier;
using ProductionManagement.Infrastructure.Persistence.TypeHandlers;

namespace ProductionManagement.Infrastructure.Persistence.Repositories;

/// <summary>
/// 取引先リポジトリ実装
/// </summary>
public class SupplierRepository : ISupplierRepository
{
    private readonly string _connectionString;

    static SupplierRepository()
    {
        SqlMapper.AddTypeHandler(new DateOnlyTypeHandler());
        SqlMapper.AddTypeHandler(new SupplierTypeTypeHandler());
    }

    public SupplierRepository(string connectionString)
    {
        _connectionString = connectionString;
    }

    public async Task SaveAsync(Supplier supplier)
    {
        const string sql = """
            INSERT INTO "取引先マスタ" (
                "取引先コード", "適用開始日", "適用停止日", "取引先名", "取引先カナ",
                "取引先区分", "郵便番号", "住所", "電話番号", "FAX番号", "担当者名"
            ) VALUES (
                @SupplierCode, @EffectiveFrom, @EffectiveTo, @SupplierName, @SupplierNameKana,
                @SupplierType::取引先区分, @PostalCode, @Address, @PhoneNumber, @FaxNumber, @ContactPerson
            )
            """;

        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.ExecuteAsync(sql, new
        {
            supplier.SupplierCode,
            supplier.EffectiveFrom,
            supplier.EffectiveTo,
            supplier.SupplierName,
            supplier.SupplierNameKana,
            SupplierType = supplier.SupplierType.GetDisplayName(),
            supplier.PostalCode,
            supplier.Address,
            supplier.PhoneNumber,
            supplier.FaxNumber,
            supplier.ContactPerson
        });
    }

    public async Task<Supplier?> FindByCodeAsync(string supplierCode)
    {
        const string sql = """
            SELECT
                "取引先コード" as SupplierCode,
                "適用開始日" as EffectiveFrom,
                "適用停止日" as EffectiveTo,
                "取引先名" as SupplierName,
                "取引先カナ" as SupplierNameKana,
                "取引先区分"::TEXT as SupplierTypeValue,
                "郵便番号" as PostalCode,
                "住所" as Address,
                "電話番号" as PhoneNumber,
                "FAX番号" as FaxNumber,
                "担当者名" as ContactPerson,
                "作成日時" as CreatedAt,
                "更新日時" as UpdatedAt
            FROM "取引先マスタ"
            WHERE "取引先コード" = @SupplierCode
            ORDER BY "適用開始日" DESC
            LIMIT 1
            """;

        await using var connection = new NpgsqlConnection(_connectionString);
        return await connection.QuerySingleOrDefaultAsync<Supplier>(sql, new { SupplierCode = supplierCode });
    }

    public async Task<Supplier?> FindByCodeAndDateAsync(string supplierCode, DateOnly baseDate)
    {
        const string sql = """
            SELECT
                "取引先コード" as SupplierCode,
                "適用開始日" as EffectiveFrom,
                "適用停止日" as EffectiveTo,
                "取引先名" as SupplierName,
                "取引先カナ" as SupplierNameKana,
                "取引先区分"::TEXT as SupplierTypeValue,
                "郵便番号" as PostalCode,
                "住所" as Address,
                "電話番号" as PhoneNumber,
                "FAX番号" as FaxNumber,
                "担当者名" as ContactPerson,
                "作成日時" as CreatedAt,
                "更新日時" as UpdatedAt
            FROM "取引先マスタ"
            WHERE "取引先コード" = @SupplierCode
              AND "適用開始日" <= @BaseDate
              AND ("適用停止日" IS NULL OR "適用停止日" > @BaseDate)
            ORDER BY "適用開始日" DESC
            LIMIT 1
            """;

        await using var connection = new NpgsqlConnection(_connectionString);
        return await connection.QuerySingleOrDefaultAsync<Supplier>(sql, new { SupplierCode = supplierCode, BaseDate = baseDate });
    }

    public async Task<IReadOnlyList<Supplier>> FindByTypeAsync(SupplierType supplierType)
    {
        const string sql = """
            SELECT
                "取引先コード" as SupplierCode,
                "適用開始日" as EffectiveFrom,
                "適用停止日" as EffectiveTo,
                "取引先名" as SupplierName,
                "取引先カナ" as SupplierNameKana,
                "取引先区分"::TEXT as SupplierTypeValue,
                "郵便番号" as PostalCode,
                "住所" as Address,
                "電話番号" as PhoneNumber,
                "FAX番号" as FaxNumber,
                "担当者名" as ContactPerson,
                "作成日時" as CreatedAt,
                "更新日時" as UpdatedAt
            FROM "取引先マスタ"
            WHERE "取引先区分" = @SupplierType::取引先区分
              AND ("適用停止日" IS NULL OR "適用停止日" > CURRENT_DATE)
            ORDER BY "取引先コード", "適用開始日" DESC
            """;

        await using var connection = new NpgsqlConnection(_connectionString);
        var result = await connection.QueryAsync<Supplier>(sql, new { SupplierType = supplierType.GetDisplayName() });
        return result.ToList();
    }

    public async Task<IReadOnlyList<Supplier>> FindAllAsync()
    {
        const string sql = """
            SELECT
                "取引先コード" as SupplierCode,
                "適用開始日" as EffectiveFrom,
                "適用停止日" as EffectiveTo,
                "取引先名" as SupplierName,
                "取引先カナ" as SupplierNameKana,
                "取引先区分"::TEXT as SupplierTypeValue,
                "郵便番号" as PostalCode,
                "住所" as Address,
                "電話番号" as PhoneNumber,
                "FAX番号" as FaxNumber,
                "担当者名" as ContactPerson,
                "作成日時" as CreatedAt,
                "更新日時" as UpdatedAt
            FROM "取引先マスタ"
            ORDER BY "取引先コード", "適用開始日" DESC
            """;

        await using var connection = new NpgsqlConnection(_connectionString);
        var result = await connection.QueryAsync<Supplier>(sql);
        return result.ToList();
    }

    public async Task DeleteAllAsync()
    {
        const string sql = """TRUNCATE TABLE "取引先マスタ" CASCADE""";

        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.ExecuteAsync(sql);
    }
}
